### ISSUE TITLE
[Question] Process frames parallelly

### ISSUE THREAD

**Issue by DaeyangCho** [nnstreamer/nnstreamer#3688]
Hi maintainers and committers.
Current tensor_filter pytorch and python3 code works per frame.
Is there any way to process frames parallelly?

Here are some example flows.
1.
![Screenshot from 2022-03-08 17-07-46](https://user-images.githubusercontent.com/93561142/157193793-8d280ab6-6de4-4c73-b14f-e301d5fab64d.png)
2.
![Screenshot from 2022-03-08 17-08-28](https://user-images.githubusercontent.com/93561142/157193843-f83877f2-ecc0-4cb4-9e00-477c52a751ae.png)

Any comments or suggestions or debugging approach would be appreciated!

**Comment by myungjoo** [comment]
Approach 1: yes, you can do it right now. when you merge them, you may use ```join``` element. However, synchronization (the order is not guaranteed) may need some care.

Approach 2: if you want this, you will need to write a subplugin or python code to do parallelism. However, in general, it is more efficient to have a "single" tensor_filter and process a single neural network in parallel unless the size of neural network is very small.

Note that, in most cases, having a single tensor_filter, processing one neural network at a time, and enable parallelism for a single instance of inference is more efficient.

**Comment by myungjoo** [comment]
@gichan-jang  (author of join): do you have an idea on Approach 1?

What do you tink about...

```
... ! tensor_filter (odd) ! p.sink_0
... ! tensor_filter (even) ! p.sink_1
join name=p ! queue min-threshold-buffers=2 ! ...
```

Refer: https://gstreamer.freedesktop.org/documentation/coreelements/queue.html?gi-language=c

**Comment by gichan-jang** [comment]
If @DaeyangCho wants output stream after tensor filter (... - odd - even - odd - even ... ) , he can use `join`.
But, If he wants merged frame (odd + even), `join` is not suitable.
@DaeyangCho, Please explain more about the odd and even frames (interlace?) and the merged frame.

**Comment by DaeyangCho** [comment]
@myungjoo Thanks for your kind reply.
For approach 1, I'm trying to figure out how to split the input stream into even and odd frames.
It would be very helpful if you could tell me how to divide it into even and odd frames.
@gichan-jang What I want is the first case, and the order is not shuffled (e.g., { 0, 1, 2, 3, 4, 5 }).

@myungjoo For approach 2, could you clarify the meaning of "enable parallelism for a single instance of inference"?
In my case, the inference time of the model for one frame is almost 400ms, and it only consumes about 30% GPU utilization.
I wonder if the parallelism we're talking about here means taking a large batch size.
If so, I wonder how I can take the batch size larger than 1 in tensor_filter for pytorch or python3.

**Comment by DaeyangCho** [comment]
@gichan-jang Isn't there something similar to join among the gstreamer plugins?

**Comment by gichan-jang** [comment]
> @gichan-jang Isn't there something similar to join among the gstreamer plugins?

As I know, there is no element in the gstreamer. I made a `join` by referring to the `input-selector`.
And if you haven't found the proper way to split the video, how about using `sensor_if`?
For more efficient operation, there must be a change in tensor_if operator, but if you want to use it now, you can try this.

```
gst-launch-1.0 \
    v4l2src ! videoscale ! videoconvert ! video/x-raw,format=RGB,width=640,height=480,framerate=10/1 ! tensor_converter ! mux.sink_0 \
    videotestsrc pattern=blink ! videoscale ! videoconvert ! video/x-raw,format=RGB,width=2,height=2,framerate=10/1 ! tensor_converter ! mux.sink_1 \
    tensor_mux name=mux ! tensor_if name=tif compared-value=A_VALUE compared-value-option=0:0:0:0,1 \
                                    supplied-value=100 operator=GE then=TENSORPICK then-option=0 else=TENSORPICK else-option=0 \
        tif.src_0 ! queue leaky=2 max-size-buffers=2 ! tensor_decoder mode=direct_video ! videoconvert ! ximagesink \
        tif.src_1 ! queue leaky=2 max-size-buffers=2 ! tensor_decoder mode=direct_video ! videoconvert ! ximagesink
```

**Comment by gichan-jang** [comment]
> @gichan-jang (author of join): do you have an idea on Approach 1?
> 
> What do you tink about...
> 
> ```
> ... ! tensor_filter (odd) ! p.sink_0
> ... ! tensor_filter (even) ! p.sink_1
> ```
> 
> ...
> 
> ```
> join name=p ! queue min-threshold-buffers=2 ! ...
> ```
> 
> (assuming that the video stream feeds frames infinitely...)
> 
> or
> 
> ```
> join name=p ! queue min-threshold-time=50000000 ! ...
> ```
> 
> (assuming that 50ms is enough latency to wait for an alternative tensor-filter)
> 
> Refer: https://gstreamer.freedesktop.org/documentation/coreelements/queue.html?gi-language=c

It would be possible if the processing time of the tensor filter did not change significantly.
And shouldn't `queue` insert before `join`?
```
 ... ! tensor_filter (odd) ! p.sink_0 \
 ... ! tensor_filter (even) ! queue min-threshold-time=${PROPER_DELAY} ! p.sink_1 \
 join name=p ! ...
```

**Comment by myungjoo** [comment]
> @myungjoo Thanks for your kind reply. For approach 1, I'm trying to figure out how to split the input stream into even and odd frames. It would be very helpful if you could tell me how to divide it into even and odd frames. @gichan-jang What I want is the first case, and the order is not shuffled (e.g., { 0, 1, 2, 3, 4, 5 }).

Idea 1-1: merge two frames into one frame with tensor_aggregator and split them into to streams with tensor_split
```
  ... ! tensor_aggregator (some options) ! tensor_split (some options) name=ts
ts.src_0 ! ....
ts.src_1 ! ...
```

Reference: https://github.com/nnstreamer/nnstreamer/tree/main/gst/nnstreamer/tensor_aggregator

> 
> @myungjoo For approach 2, could you clarify the meaning of "enable parallelism for a single instance of inference"? In my case, the inference time of the model for one frame is almost 400ms, and it only consumes about 30% GPU utilization. I wonder if the parallelism we're talking about here means taking a large batch size. If so, I wonder how I can take the batch size larger than 1 in tensor_filter for pytorch or python3.

Before doing anything, you must profile why the utilization is so low.
If 70% of the time is used by memory transfer between GPU and main memory, you are already bottlenecked and trying multiple instances of inferences will only DETERIORATE the performance.
And if you really think that (I doubt it!) you can improve the performance by calling multiple instances of inferences (e.g., invoking multiple tensor_filter), you should try to increase the number of threads or GPU tasks in parallel for a single instance of input buffer (in a single instance of tensor_filter).

Your approach will ONLY work IF you have multiple processors that CANNOT be shared or collaborate for a single neural network model.

**Comment by DaeyangCho** [comment]
@gichan-jang @myungjoo 답변 주셔서 정말 감사합니다.
join과 tensor_if, tensor_aggregator 등 말씀해주신 것들을 통해서 생각했던 것들을 테스트 해 볼 수 있을 것 같습니다.
상세한 예제 또한 큰 도움이 되었습니다.

회사에서 python으로 짜여진 여러 모델들을 빠르게 nnstreamer에서 돌려보는 것을 요청하여서, tensor_filter python3로 돌려보는 중입니다.
이 과정에서 fps가 낮게 나와서, 빠른 차선책이 있나해서 여쭤보았습니다. (Python 코드 안에 cpu를 비효율적으로 사용하는 cv 코드들이 있습니다.)
앞으로, 말씀 주신대로, 모델들을 프로파일링하고, torchscript등으로 컴파일하고, CV쪽 전처리, 후처리도 cpp로 넘어가려고 합니다.
NNStreamer라는 좋을 프레임워크를 제공해주셔서 감사합니다!

**Comment by myungjoo** [comment]
실제 상업용으로 쓸 모델이라면 Python 없이 곧바로 Tensorflow, Pytorch 등으로 Inference할 수 있어야 합니다.
그렇게 Model deploy를 하시고 (media 처리이겠죠?) CV로 짠 부분을 왠만하면 GStreamer pipeline으로 표현하시면 됩니다.
Python/CV 코드들이 있는 상태는 어디까지나 연구/Prototyping용이지, 성능 최적화가 수반되는 Deploy용으로는 곤란하지 않을까요? ^^

**Comment by DaeyangCho** [comment]
넵넵 맞습니다. Python은 기획팀에서 느낌을 보고 싶어해서 프로토타입용으로 간단히 만든 것이었습니다.
최적화 잘해서 NNStreamer에 붙여보도록 하겠습니다 ㅎㅎ 감사합니다!
